// script-bundler/Gruntfile.js
/*
 * script-bundler
 * http://github.com/petervyvey/script-bundler
 *
 * Copyright (c) 2014 Peter Vyvey
 * Licensed under the MIT license.
 */

'use strict';

module.exports = function (grunt) {

    require('load-grunt-tasks')(grunt);
    grunt.loadTasks('tasks');

    // Project configuration.
    grunt.initConfig({
        jshint: {
            all: [
                'Gruntfile.js',
                'tasks/*.js',
                '<%= nodeunit.tests %>'
            ],
            options: {
                jshintrc: '.jshintrc',
                reporter: require('jshint-stylish')
            }
        },

        // Before generating any new files, remove any previously-created files.
        clean: {
            tests: ['tmp']
        },

        // Configuration to be run (and then tested).
        script_bundler: {
            default_options: {
                options: {
                },
                files: {
                    'tmp': ['test/fixtures/JavaScript.json']
                }
            }
        },

        // Unit tests.
        nodeunit: {
            tests: ['test/*_test.js']
        }

    });

    // Whenever the "test" task is run, first clean the "tmp" dir, then run this
    // plugin's task(s), then test the result.
    grunt.registerTask('test', ['clean', 'script_bundler', 'nodeunit']);

    grunt.registerTask('release', ['clean', 'script_bundler']);
    grunt.registerTask('default', ['jshint', 'test']);

};
